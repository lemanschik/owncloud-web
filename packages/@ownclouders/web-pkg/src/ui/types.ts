import { IconFillType } from 'design-system/src/helpers'

export type ViewMode = {
  name: string
  label: string
  icon: {
    name: string
    fillType: IconFillType
  }
}
