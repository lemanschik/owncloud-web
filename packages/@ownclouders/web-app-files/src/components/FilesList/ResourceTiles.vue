<template>
  <div>
    <div v-if="sortFields.length" class="oc-tile-sorting oc-border-b oc-mb-m oc-pb-s">
      <span class="oc-mr-xs" v-text="$gettext('Sort by: ')" />
      <oc-button id="oc-tiles-sort-btn" appearance="raw" gap-size="none">
        <span v-text="$gettext(currentSortField.label)" />
        <oc-icon name="arrow-down-s" />
      </oc-button>
      <oc-drop
        ref="sortDrop"
        toggle="#oc-tiles-sort-btn"
        class="oc-tiles-sort-drop"
        mode="click"
        padding-size="small"
        close-on-click
      >
        <oc-list class="oc-tiles-sort-list">
          <li v-for="(field, index) in sortFields" :key="index" class="oc-my-xs">
            <oc-button
              justify-content="space-between"
              class="oc-tiles-sort-list-item oc-p-s oc-width-1-1"
              :class="{
                'oc-background-primary-gradient': isSortFieldSelected(field),
                selected: isSortFieldSelected(field)
              }"
              :appearance="isSortFieldSelected(field) ? 'raw-inverse' : 'raw'"
              :variation="isSortFieldSelected(field) ? 'primary' : 'passive'"
              @click="selectSorting(field)"
            >
              <span v-text="$gettext(field.label)" />
              <oc-icon v-if="isSortFieldSelected(field)" name="check" variation="inherit" />
            </oc-button>
          </li>
        </oc-list>
      </oc-drop>
    </div>
    <oc-list class="oc-tiles oc-flex" :class="resizable ? 'resizableTiles' : ''">
      <li v-for="resource in data" :key="resource.id" class="oc-tiles-item has-item-context-menu">
        <oc-tile
          :ref="(el) => (tileRefs.tiles[resource.id] = el)"
          :resource="resource"
          :resource-route="getRoute(resource)"
          :is-resource-selected="isResourceSelected(resource)"
          :is-extension-displayed="areFileExtensionsShown"
          :resource-icon-size="resourceIconSize"
          :draggable="dragDrop"
          @vue:mounted="
            $emit('rowMounted', resource, tileRefs.tiles[resource.id], ImageDimension.Tile)
          "
          @contextmenu="showContextMenu($event, resource.id, tileRefs.tiles[resource.id])"
          @click="emitTileClick(resource)"
          @dragstart="dragStart(resource, $event)"
          @dragenter.prevent="setDropStyling(resource, false, $event)"
          @dragleave.prevent="setDropStyling(resource, true, $event)"
          @drop="fileDropped(resource, $event)"
          @dragover="$event.preventDefault()"
        >
          <template #selection>
            <oc-checkbox
              :label="getResourceCheckboxLabel(resource)"
              :hide-label="true"
              size="large"
              class="oc-flex-inline oc-p-s"
              :model-value="isResourceSelected(resource)"
              @click.stop.prevent="toggleSelection(resource)"
            />
          </template>
          <template #imageField>
            <slot name="image" :resource="resource" />
          </template>
          <template #actions>
            <slot name="actions" :resource="resource" />
          </template>
          <template #contextMenu>
            <context-menu-quick-action
              :ref="(el) => (tileRefs.dropBtns[resource.id] = el)"
              :item="resource"
              class="resource-tiles-btn-action-dropdown"
              @quick-action-clicked="showContextMenuOnBtnClick($event, resource, resource.id)"
            >
              <template #contextMenu>
                <slot name="contextMenuActions" :resource="resource" />
              </template>
            </context-menu-quick-action>
          </template>
        </oc-tile>
      </li>
    </oc-list>
    <Teleport v-if="dragItem" to="body">
      <oc-ghost-element ref="ghostElementRef" :preview-items="[dragItem, ...dragSelection]" />
    </Teleport>
    <div class="oc-tiles-footer">
      <slot name="footer" />
    </div>
  </div>
</template>

<script lang="ts">
import { onBeforeUpdate, defineComponent, nextTick, PropType, computed, ref, unref } from 'vue'
import { useGettext } from 'vue3-gettext'
import { Resource, SpaceResource } from 'web-client'
import { useStore, SortDir, SortField } from 'web-pkg/src/composables'
import { ImageDimension } from 'web-pkg/src/constants'
import { createFileRouteOptions } from 'web-pkg/src/helpers/router'
import { displayPositionedDropdown } from 'web-pkg/src/helpers/contextMenuDropdown'
import { createLocationSpaces } from 'web-app-files/src/router'
import ContextMenuQuickAction from 'web-pkg/src/components/ContextActions/ContextMenuQuickAction.vue'
import { ViewModeConstants } from 'web-app-files/src/composables'

// Constants should match what is being used in OcTable/ResourceTable
// Alignment regarding naming would be an API-breaking change and can
// Be done at a later point in time?
import { useResourceRouteResolver } from '../../composables/filesList'

export default defineComponent({
  name: 'ResourceTiles',
  components: { ContextMenuQuickAction },
  props: {
    /**
     * Array of resources (spaces, folders, files) to be displayed as tiles
     */
    data: {
      type: Array as PropType<Resource[]>,
      default: () => []
    },
    resizable: {
      type: Boolean,
      default: false
    },
    selectedIds: {
      type: Array,
      default: () => []
    },
    targetRouteCallback: {
      type: Function,
      required: false,
      default: undefined
    },
    space: {
      type: Object as PropType<SpaceResource>,
      required: false,
      default: null
    },
    sortFields: {
      type: Array as PropType<SortField[]>,
      default: () => []
    },
    sortBy: {
      type: String,
      required: false,
      default: undefined
    },
    sortDir: {
      type: String,
      required: false,
      default: undefined,
      validator: (value: string) => {
        return (
          value === undefined || [SortDir.Asc.toString(), SortDir.Desc.toString()].includes(value)
        )
      }
    },
    viewSize: {
      type: Number,
      required: false,
      default: ViewModeConstants.tilesSizeDefault
    },
    dragDrop: {
      type: Boolean,
      default: false
    }
  },
  emits: ['fileClick', 'fileDropped', 'rowMounted', 'sort', 'update:selectedIds'],
  setup(props, context) {
    const store = useStore()
    const { $gettext } = useGettext()

    const areFileExtensionsShown = computed(() => store.state.Files.areFileExtensionsShown)

    const dragItem = ref()
    const ghostElementRef = ref()
    const tileRefs = ref({
      tiles: [],
      dropBtns: []
    })

    const spaces = computed(() => {
      return store.getters['runtime/spaces/spaces']
    })

    const resourceRouteResolver = useResourceRouteResolver(
      {
        space: ref(props.space),
        spaces,
        targetRouteCallback: computed(() => props.targetRouteCallback)
      },
      context
    )

    const getRoute = (resource) => {
      if (resource.type === 'space') {
        return resource.disabled
          ? { path: '#' }
          : createLocationSpaces(
              'files-spaces-generic',
              createFileRouteOptions(resource as SpaceResource, {
                path: '',
                fileId: resource.fileId
              })
            )
      }
      if (resource.type === 'folder') {
        return resourceRouteResolver.createFolderLink({
          path: resource.path,
          fileId: resource.fileId,
          resource: resource
        })
      }
      return { path: '' }
    }

    const emitTileClick = (resource) => {
      if (resource.disabled && resource.type === 'space') {
        store.dispatch('showMessage', {
          title: $gettext('Disabled spaces cannot be entered'),
          status: 'warning'
        })
      }
      if (resource.type !== 'space' && resource.type !== 'folder') {
        resourceRouteResolver.createFileAction(resource)
      }
    }

    const showContextMenuOnBtnClick = (data, item, index) => {
      const { dropdown, event } = data
      if (dropdown?.tippy === undefined) {
        return
      }
      displayPositionedDropdown(dropdown.tippy, event, unref(tileRefs).dropBtns[index])
    }

    const showContextMenu = (event, index, reference) => {
      event.preventDefault()
      const drop = unref(tileRefs).tiles[index]?.$el.getElementsByClassName(
        'resource-tiles-btn-action-dropdown'
      )[0]

      if (drop === undefined) {
        return
      }
      displayPositionedDropdown(drop._tippy, event, reference)
    }

    const isResourceSelected = (resource) => {
      return props.selectedIds.includes(resource.id)
    }

    const toggleSelection = (resource) => {
      const selectedIds = !isResourceSelected(resource)
        ? [...props.selectedIds, resource.id]
        : props.selectedIds.filter((id) => id !== resource.id)
      context.emit('update:selectedIds', selectedIds)
    }

    const getResourceCheckboxLabel = (resource) => {
      switch (resource.type) {
        case 'folder':
          return $gettext('Select folder')
        case 'space':
          return $gettext('Select space')
        default:
          return $gettext('Select file')
      }
    }

    const currentSortField = computed(() => {
      return (
        props.sortFields.find((o) => o.name === props.sortBy && o.sortDir === props.sortDir) ||
        props.sortFields[0]
      )
    })
    const selectSorting = (field) => {
      context.emit('sort', { sortBy: field.name, sortDir: field.sortDir })
    }
    const isSortFieldSelected = (field) => {
      return unref(currentSortField) === field
    }

    const resourceIconSize = computed(() => {
      const sizeMap = {
        1: 'xlarge',
        2: 'xlarge',
        3: 'xxlarge',
        4: 'xxlarge',
        5: 'xxxlarge',
        6: 'xxxlarge'
      }
      return sizeMap[props.viewSize] ?? 'xlarge'
    })

    onBeforeUpdate(() => {
      tileRefs.value = {
        tiles: [],
        dropBtns: []
      }
    })

    const setDropStyling = (resource, leaving, event) => {
      const hasFilePayload = (event.dataTransfer?.types || []).some((e) => e === 'Files')
      if (
        hasFilePayload ||
        event.currentTarget?.contains(event.relatedTarget) ||
        props.selectedIds.includes(resource.id) ||
        resource.type !== 'folder'
      ) {
        return
      }
      const el = unref(tileRefs).tiles[resource.id]
      if (leaving) {
        el.$el.classList.remove('oc-tiles-item-drop-highlight')
        return
      }
      el.$el.classList.add('oc-tiles-item-drop-highlight')
    }
    const dragSelection = computed(() => {
      return props.selectedIds.filter((id) => id !== unref(dragItem).id)
    })
    const setDragItem = async (item, event) => {
      dragItem.value = item
      await nextTick()
      unref(ghostElementRef).$el.ariaHidden = 'true'
      unref(ghostElementRef).$el.style.left = '-99999px'
      unref(ghostElementRef).$el.style.top = '-99999px'
      event.dataTransfer.setDragImage(unref(ghostElementRef).$el, 0, 0)
      event.dataTransfer.dropEffect = 'move'
      event.dataTransfer.effectAllowed = 'move'
    }
    const dragStart = async (resource, event) => {
      if (!isResourceSelected(resource)) {
        toggleSelection(resource)
      }
      await setDragItem(resource, event)
    }

    const fileDropped = (resource, event) => {
      const hasFilePayload = (event.dataTransfer.types || []).some((e) => e === 'Files')
      if (hasFilePayload) {
        return
      }
      dragItem.value = null
      setDropStyling(resource, true, event)
      context.emit('fileDropped', resource.id)
    }

    return {
      areFileExtensionsShown,
      emitTileClick,
      getRoute,
      showContextMenuOnBtnClick,
      showContextMenu,
      tileRefs,
      isResourceSelected,
      toggleSelection,
      getResourceCheckboxLabel,
      selectSorting,
      isSortFieldSelected,
      currentSortField,
      resourceIconSize,
      ghostElementRef,
      dragItem,
      dragStart,
      dragSelection,
      fileDropped,
      setDropStyling
    }
  },
  data() {
    return {
      ImageDimension
    }
  }
})
</script>

<style lang="scss" scoped>
.oc-tiles {
  column-gap: 1rem;
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(auto, var(--oc-size-tiles-default)));
  justify-content: flex-start;
  row-gap: 1rem;

  &.resizableTiles {
    grid-template-columns: repeat(auto-fill, minmax(auto, var(--oc-size-tiles-resize-step)));
  }

  @media only screen and (max-width: 640px) {
    grid-template-columns: 80%;
    justify-content: center;
    padding: var(--oc-space-medium) 0;

    &.resizableTiles {
      grid-template-columns: 80%;
    }
  }

  &-item-drop-highlight {
    background-color: var(--oc-color-input-border) !important;
  }

  &-footer {
    color: var(--oc-color-text-muted);
    font-size: var(--oc-font-size-default);
    line-height: 1.4;
    padding: var(--oc-space-xsmall);
  }

  &-sort-drop {
    width: 200px;
  }

  &-sort-list {
    &-item {
      &:hover,
      &:focus {
        background-color: var(--oc-color-background-hover);
        text-decoration: none;
      }
    }
  }
}
</style>
