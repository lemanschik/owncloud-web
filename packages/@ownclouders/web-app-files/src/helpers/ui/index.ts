export * as fileList from './filesList'
export * as resourceTable from './resourceTable'
export * as resourceTiles from './resourceTiles'
