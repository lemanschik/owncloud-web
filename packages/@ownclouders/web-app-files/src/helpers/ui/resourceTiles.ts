import { SortDir, SortField } from 'web-pkg/src/composables'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

export const sortFields: SortField[] = [
  {
    label: $gettext('A-Z'),
    name: 'name',
    sortable: true,
    sortDir: SortDir.Asc
  },
  {
    label: $gettext('Z-A'),
    name: 'name',
    sortable: true,
    sortDir: SortDir.Desc
  },
  {
    label: $gettext('Newest'),
    name: 'mdate',
    sortable: (date) => new Date(date).valueOf(),
    sortDir: SortDir.Desc
  },
  {
    label: $gettext('Oldest'),
    name: 'mdate',
    sortable: (date) => new Date(date).valueOf(),
    sortDir: SortDir.Asc
  },
  {
    label: $gettext('Largest'),
    name: 'size',
    sortable: true,
    sortDir: SortDir.Desc
  },
  {
    label: $gettext('Smallest'),
    name: 'size',
    sortable: true,
    sortDir: SortDir.Asc
  }
]

export const determineSortFields = (firstResource): SortField[] => {
  if (!firstResource) {
    return []
  }

  return sortFields.filter((field) =>
    Object.prototype.hasOwnProperty.call(firstResource, field.name)
  )
}
