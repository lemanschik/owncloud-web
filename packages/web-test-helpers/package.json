{
  "name": "web-test-helpers",
  "private": true,
  "main": "src/index.ts",
  "peerDependencies": {
    "@casl/ability": "^6.3.3",
    "@casl/vue": "^2.2.1",
    "axios": "0.27.2",
    "design-system": "@ownclouders/design-system",
    "vue3-gettext": "^2.3.3",
    "vue-router": "4.1.6",
    "vuex": "4.1.0"
  }
}
