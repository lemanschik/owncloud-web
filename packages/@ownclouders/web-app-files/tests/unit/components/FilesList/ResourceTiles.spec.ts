import { createStore, defaultPlugins, defaultStoreMockOptions, mount } from 'web-test-helpers'
import ResourceTiles from 'web-app-files/src/components/FilesList/ResourceTiles.vue'
import { sortFields } from 'web-app-files/src/helpers/ui/resourceTiles'
import { Resource } from 'web-client'
import { mock } from 'jest-mock-extended'

const spacesResources = [
  {
    id: '1',
    name: 'Space 1',
    path: '',
    type: 'space',
    isFolder: true,
    getDriveAliasAndItem: () => '1'
  },
  {
    id: '2',
    name: 'Space 2',
    path: '',
    type: 'space',
    isFolder: true,
    getDriveAliasAndItem: () => '2'
  }
]

describe('ResourceTiles component', () => {
  it('renders an array of spaces correctly', () => {
    const { wrapper } = getWrapper({ data: spacesResources })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a footer slot', () => {
    const { wrapper } = getWrapper({}, { footer: 'Hello, ResourceTiles footer!' })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('emits fileClick event upon click on tile', async () => {
    const { wrapper } = getWrapper({ data: spacesResources })
    await wrapper.find('oc-tile').trigger('click')
    expect(wrapper.emitted('click')).toBeTruthy()
  })

  it('emits update:selectedIds event on resource selection and sets the selection', () => {
    const { wrapper } = getWrapper({ data: spacesResources, selectedIds: [spacesResources[0].id] })
    wrapper.vm.toggleSelection(spacesResources[0])
    expect(wrapper.find('oc-tile').attributes()['is-resource-selected']).toEqual('true')
    expect(wrapper.emitted('update:selectedIds')).toBeTruthy()
  })

  describe('sorting', () => {
    it('renders the label of the first sort field as default', () => {
      const { wrapper } = getWrapper({ sortFields })
      expect(wrapper.find('#oc-tiles-sort-btn').text()).toEqual(sortFields[0].label)
    })
    it('renders the label of the current sort field as default', () => {
      const sortField = sortFields[2]
      const { wrapper } = getWrapper({
        sortFields,
        sortBy: sortField.name,
        sortDir: sortField.sortDir
      })
      expect(wrapper.find('#oc-tiles-sort-btn').text()).toEqual(sortField.label)
    })
    it('emits the "sort"-event', async () => {
      const index = 2
      const { wrapper } = getWrapper({ sortFields })
      await wrapper.findAll('.oc-tiles-sort-list-item').at(index).trigger('click')
      expect(wrapper.emitted('sort')).toBeTruthy()
      expect(wrapper.emitted('sort')[0][0]).toEqual({
        sortBy: sortFields[index].name,
        sortDir: sortFields[index].sortDir
      })
    })
    describe('drag and drop', () => {
      it('emits the "update:selectedIds"-event on drag start', async () => {
        const { wrapper } = getWrapper()
        wrapper.vm.dragItem = mock<Resource>()
        await wrapper.vm.$nextTick()
        ;(wrapper.vm.$refs.ghostElementRef as any).$el = { style: {} }
        wrapper.vm.dragStart(mock<Resource>(), { dataTransfer: { setDragImage: jest.fn() } })
        expect(wrapper.emitted('update:selectedIds')).toBeDefined()
      })
      it('emits the "fileDropped"-event on resource drop', () => {
        const { wrapper } = getWrapper()
        wrapper.vm.fileDropped(mock<Resource>(), { dataTransfer: {} })
        expect(wrapper.emitted('fileDropped')).toBeDefined()
      })
    })
  })

  it.each([
    { viewSize: 1, expected: 'xlarge' },
    { viewSize: 2, expected: 'xlarge' },
    { viewSize: 3, expected: 'xxlarge' },
    { viewSize: 4, expected: 'xxlarge' },
    { viewSize: 5, expected: 'xxxlarge' },
    { viewSize: 6, expected: 'xxxlarge' }
  ])('passes the "viewSize" to the OcTile component', (data) => {
    const { wrapper } = getWrapper({ data: spacesResources, viewSize: data.viewSize })
    expect(wrapper.find('oc-tile').attributes()['resource-icon-size']).toEqual(data.expected)
  })

  function getWrapper(props = {}, slots = {}) {
    const storeOptions = defaultStoreMockOptions
    const store = createStore(storeOptions)

    return {
      wrapper: mount(ResourceTiles, {
        props: {
          ...props
        },
        slots: {
          ...slots
        },
        global: {
          plugins: [...defaultPlugins({ designSystem: false }), store]
        }
      })
    }
  }
})
