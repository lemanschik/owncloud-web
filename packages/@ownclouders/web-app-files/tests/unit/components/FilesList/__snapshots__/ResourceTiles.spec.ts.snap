// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceTiles component renders a footer slot 1`] = `
<div>
  <!--v-if-->
  <oc-list class="oc-tiles oc-flex"></oc-list>
  <!--v-if-->
  <div class="oc-tiles-footer">Hello, ResourceTiles footer!</div>
</div>
`;

exports[`ResourceTiles component renders an array of spaces correctly 1`] = `
<div>
  <!--v-if-->
  <oc-list class="oc-tiles oc-flex">
    <li class="oc-tiles-item has-item-context-menu">
      <oc-tile draggable="false" is-resource-selected="false" resource="[object Object]" resource-icon-size="xlarge" resource-route="[object Object]"></oc-tile>
    </li>
    <li class="oc-tiles-item has-item-context-menu">
      <oc-tile draggable="false" is-resource-selected="false" resource="[object Object]" resource-icon-size="xlarge" resource-route="[object Object]"></oc-tile>
    </li>
  </oc-list>
  <!--v-if-->
  <div class="oc-tiles-footer"></div>
</div>
`;
